tools/aws/create-spot-cluster.py

```diff
@@ -0,0 +1,200 @@
+    pp.pprint(resp)
+    
+    # get request ids
+    requests = resp['SpotInstanceRequests']
+    assert len(requests) == INSTANCE_COUNT
+    request_ids[conf['region']] = [r['SpotInstanceRequestId'] for r in requests]
+
+  print('\n{:=^50}'.format('WAIT FOR INSTANCES'))
+  instance_ips = {}
+  for region, request_id in request_ids.items():
+    print(region)
+    ips = []
+    retries = 0
+    while len(ips) < INSTANCE_COUNT:
+      retries += 1
+      assert retries < MAX_RETRIES
+
+      wait_time = REFRESH_WAIT_SCALE ** retries # exponential backoff
+      print("Waiting " + str(wait_time) + " seconds...")
+      time.sleep(wait_time)
+
+      print("Trying to fetch " + str(INSTANCE_COUNT) + " instances...")
+      ips = get_instance_ips(region, request_id)
+      print("Got " + str(len(ips)))
+
+    assert len(ips) == INSTANCE_COUNT
+    instance_ips[region] = ips
+
+  print('\nIp addresses:')
+  pp.pprint(instance_ips)
+
+  print ('\n{:=^50}'.format('INSTALL DOCKER'))
+  time.sleep(SSH_DELAY_SECONDS)
+  install_docker(instance_ips)
+
+  print('\nIp addresses:')
+  pp.pprint(instance_ips)
+
+
+def request_spot_instances(conf):
+  ec2 = boto3.client('ec2', region_name=conf['region'])
+  return ec2.request_spot_instances(
+    BlockDurationMinutes=SPOT_DURATION,
+    InstanceCount=INSTANCE_COUNT,
+    LaunchSpecification={
+      'SecurityGroups': [
+        'ssh',
+        'slog',
+      ],
+      'ImageId': conf['image_id'],
+      'InstanceType': conf['instance_type'],
+      'KeyName': conf['key_name'],
+      'Placement': {
+        'AvailabilityZone': conf['availability_zone'],
+      },
+    }
+  )
+
+
+def get_instance_ips(region, ids):
+  ec2 = boto3.client('ec2', region_name=region)
+  resp = ec2.describe_spot_instance_requests(
+    SpotInstanceRequestIds = ids
+  )
+  # TODO: surface if request closed
+  instance_ids = []
+  for spot_req in resp['SpotInstanceRequests']:
+    if 'InstanceId' in spot_req:
+      instance_ids.append(spot_req['InstanceId'])
+    else:
+      print('\nDelayed spot request:')
+      pp.pprint(spot_req)
+
+  if not instance_ids:
+    return []
+
+  ips = []
+  resp = ec2.describe_instances(
+    Filters = [
+      {
+        'Name': 'instance-state-name',
+        'Values' : ['running']
+      }
+    ],
+    InstanceIds = instance_ids,
+  )
+  for r in resp['Reservations']:
+    for i in r['Instances']:
+      if 'PublicIpAddress' in i:
+        ips.append(i['PublicIpAddress'])
+      else:
+        print('\nInstance without public ip:')
+        pp.pprint(i)
+  return ips
+
+
+def install_docker(instance_ips):
+  commands = []
+  for _, ips in instance_ips.items():
+    for ip in ips:
+      command = """ \
+        {} ubuntu@{} "{}" &\
+        """.format(SSH_COMMAND, ip, INSTALL_DOCKER_COMMAND).strip()
+      commands.append(command)
+  combined_command = "\n".join(commands) + " wait"
+  print(combined_command)
+
+  # TODO: redirect output?
+  os.system(combined_command)
```

Comment:
For more advanced features like output redirecting, you should use the `subprocess.Popen` class